extension/wasm_src/src/lib.rs

```diff
@@ -1,62 +1,78 @@
-#[cfg(cargo_web)]
-#[macro_use]
-extern crate stdweb;
-
-#[cfg(cargo_web)]
-use stdweb::js_export;
-
-use std::{cell::RefCell, rc::Rc};
-use stdweb::{traits::*, unstable::TryFrom, web};
-
-mod util;
+use railroad::RailroadNode;
+use wasm_bindgen::prelude::*;
+
+// When the `wee_alloc` feature is enabled, use `wee_alloc` as the global
+// allocator.
+#[cfg(feature = "wee_alloc")]
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(js_namespace = console)]
+    fn log(s: &str);
+}
 
-#[cfg_attr(cargo_web, js_export)]
-pub fn load() {
-    if let Err(err) = priv_load() {
-        stdweb::console!(error, format!("{:?}", err))
-    }
+#[allow(dead_code)]
+mod built_info {
+    include!(concat!(env!("OUT_DIR"), "/built.rs"));
 }
 
-#[derive(Debug)]
-pub enum Error {
-    SynParsing(syn::Error),
-    Syntax(web::error::SyntaxError),
-    InvalidCharacter(web::error::InvalidCharacterError),
-    ReplaceChild(String), // ReplaceChildError is private for some reason in stdweb 0.4.20
-    NodeNotFound,
-    #[cfg(feature = "webextension")]
-    UnexpectedType,
+#[wasm_bindgen(start)]
+pub fn start() {
+    #[cfg(feature = "console_error_panic_hook")]
+    set_panic_hook();
+
+    log(&format!(
+        "macro_railroad_ext built {} on {} using {}",
+        built_info::BUILT_TIME_UTC,
+        built_info::RUSTC_VERSION,
+        built_info::DEPENDENCIES_STR
+    ));
 }
 
-impl From<syn::Error> for Error {
-    fn from(e: syn::Error) -> Self {
-        Error::SynParsing(e)
-    }
+#[wasm_bindgen(js_name = getRailroadDefaultCss)]
+pub fn get_railroad_default_css() -> JsValue {
+    railroad::DEFAULT_CSS.into()
 }
 
-impl From<web::error::SyntaxError> for Error {
-    fn from(e: web::error::SyntaxError) -> Self {
-        Error::Syntax(e)
-    }
+#[wasm_bindgen(js_name = getRailroadDigramCss)]
+pub fn get_railroad_digram_css() -> JsValue {
+    macro_railroad::diagram::CSS.into()
 }
 
-impl From<web::error::InvalidCharacterError> for Error {
-    fn from(e: web::error::InvalidCharacterError) -> Self {
-        Error::InvalidCharacter(e)
-    }
+#[derive(Debug)]
+#[wasm_bindgen]
+pub struct DiagramOptions {
+    #[wasm_bindgen(js_name = hideInternal)]
+    pub hide_internal: bool,
+    #[wasm_bindgen(js_name = keepGroups)]
+    pub keep_groups: bool,
+    #[wasm_bindgen(js_name = foldCommonTails)]
+    pub foldcommontails: bool,
+    #[wasm_bindgen(js_name = showLegend)]
+    pub show_legend: bool,
 }
 
-type Result<T> = std::result::Result<T, Error>;
+#[wasm_bindgen]
+pub struct Diagram {
+    #[wasm_bindgen(readonly)]
+    pub width: i64,
+    svg: String,
```

Comment:
The `svg` is also `readonly` from the JS-side, isn't it?

Reply:
Actually, we can't make `svg` pub since `wasm-bindgen` require each public field to be `Copy`.  See https://github.com/rustwasm/wasm-bindgen/issues/1985 and https://github.com/rustwasm/wasm-bindgen/issues/2775.

I'll add some comments about this.